clipper_admin/clipper_admin/deployers/mxnet.py

```diff
@@ -0,0 +1,212 @@
+    registry : str, optional
+        The Docker container registry to push the freshly built model to. Note
+        that if you are running Clipper on Kubernetes, this registry must be accesible
+        to the Kubernetes cluster in order to fetch the container from the registry.
+    base_image : str, optional
+        The base Docker image to build the new model image from. This
+        image should contain all code necessary to run a Clipper model
+        container RPC client.
+    num_replicas : int, optional
+        The number of replicas of the model to create. The number of replicas
+        for a model can be changed at any time with
+        :py:meth:`clipper.ClipperConnection.set_num_replicas`.
+    """
+
+    clipper_conn.register_application(name, input_type, default_output,
+                                      slo_micros)
+    deploy_mxnet_model(clipper_conn, name, version, input_type, func,
+                       mxnet_model, mxnet_data_shapes, base_image, labels,
+                       registry, num_replicas)
+
+    clipper_conn.link_model_to_app(name, name)
+
+
+def deploy_mxnet_model(
+        clipper_conn,
+        name,
+        version,
+        input_type,
+        func,
+        mxnet_model,
+        mxnet_data_shapes,
+        base_image="clipper/mxnet-container:{}".format(__version__),
+        labels=None,
+        registry=None,
+        num_replicas=1):
+    """Deploy a Python function with a MXNet model.
+    Parameters
+    ----------
+    clipper_conn : :py:meth:`clipper_admin.ClipperConnection`
+        A ``ClipperConnection`` object connected to a running Clipper cluster.
+    name : str
+        The name to be assigned to both the registered application and deployed model.
+    version : str
+        The version to assign this model. Versions must be unique on a per-model
+        basis, but may be re-used across different models.
+    input_type : str
+        The input_type to be associated with the registered app and deployed model.
+        One of "integers", "floats", "doubles", "bytes", or "strings".
+    func : function
+        The prediction function. Any state associated with the function will be
+        captured via closure capture and pickled with Cloudpickle.
+    mxnet_model : mxnet model object
+        The MXNet model to save.
+    mxnet_data_shapes : list(int)
+        List of integers representing the dimensions of data used for model prediction.
+        Required because loading serialized MXNet models involves binding, which requires
+        the shape of the data used to train the model.
+        https://mxnet.incubator.apache.org/api/python/module.html#mxnet.module.BaseModule.bind
+
+        NOTE: Clipper may provide the model with variable size input batches. Because MXNet can't
+              handle variable size input batches, we recommend setting batch size for input data
+              to 1, or dynamically reshaping the model with every prediction based on the current
```

Comment:
Can we do this dynamic reshaping automatically for the user?

Reply:
@dcrankshaw and I had a conversation about this. Because MXNet cannot handle variable size batches, but Clipper varies batch size for optimization purposes, it would just be easier to have the user input 1 as the batch size for all MXNet models. However, I don't think we thought about automatically reshaping according to what Clipper has been doing -- that's a good thought. 
How difficult would it be to get the input batch size before each prediction? 
If it's not that difficult, maybe we can reshape the existing model according to the next batch size (and next input dimensions).

Reply:
We can obtain the batch size directly from the `inputs` parameter to the `predict` method as `batch_size = len(inputs)`. We should then be able to use this to reshape before evaluating the model, right?

Reply:
I think we would also have to get the dimensions of the data itself, e.g. (1,3,224,224). Will we will be able to read these dimensions from the `inputs` parameter?

Reply:
Ah, I see. Each input is a numpy array of datatype `float`. `np.array(inputs).shape` should do it

Reply:
In that case, I think we should be able to do it. There's a reshape() function that can reshape the model. @dcrankshaw am I missing something here? Or will this work?

Reply:
I'm confident that we'll be able to do this. Currently, the user specifies a data shape at deployment as a tuple consisting of:

1. Batch size (`deployed_shape[0]`)
2. Item dimensions (`deployed_shape[1:]`)

We can therefore compute the batch size from the inputs array using the method described above: `batch_size = len(inputs)` and set the model data shape such that 
`new_shape[0] = batch_size` 
`new_shape[1:] = deployed_shape[1:]`

Can you try this?

Reply:
The big question in my mind was about how slow re-shaping the model is. It might be an expensive operation to do for every batch. For the sake of simplicity, let's get do the automatic reshaping for the user for now. We can always optimize this later (or people can write their own model containers).

Reply:
I think it'd be best to reshape the model in the `predict` function, based off the input data passed into the function. Does the user write his own `predict` function? It seems that in the integration test we wrote our own `predict` function and called it after loading a serialized model.

Reply:
Hmm @chester-leung you raise a good point. I wonder if the right thing to do is let the user do the model re-shaping if they want from within the predict function. Automatically re-shaping for them could be confusing. That lets them choose whether to accept the overhead of reshaping in order to make batch predictions, or whether to make single predictions in a loop.

@Corey-Zumar what do you think? The idea of modifying the model internally without the users knowledge seems like a potential source of confusion.

Reply:
@dcrankshaw Relying on user specification of reshaping necessitates the extra "batch_size" flag specification to avoid incorrect behavior. Automatically handling it provides the benefits of adaptive batching and reduces the likelihood of user error. @chester-leung can you time the reshaping procedure for multiple iterations (~200) with the integration test model? If the overhead is low, I am strongly in favor of reshaping automatically.

Reply:
Do we write a `predict` function for the user? Or does the user have to write it himself? If he writes it himself similar to how we wrote it in the integration test, i.e. with data as the input, I think there's no way for us to automatically reshape -- we need the input data to know its dimensions. Since the input data is only passed into the `predict` function, and the user writes that function himself, we cannot automatically reshape.

If I'm mistaken, @Corey-Zumar do we have 200 sample data lying around? I could just reshape the model 200 times to the same initial dimensions (basically call `reshape` but not actually do any reshaping), but I'm thinking that there could be more overhead if we drastically change the model's dimensions each time we call `predict`.

Reply:
If the user writes the `predict` function themselves, we will obtain the deserialized prediction function in the container as `self.predict_fn`. We also obtain the deserialized model as `self.model`. When we receive an input, we can then mutate the model before calling `self.predict(model, input)`.

I'd like to measure the reshaping overhead. Here's the procedure:

1. Fix an initial data shape (you can use the same one that's specified in the integration test)
2. 
```
for batch_size in range(1, 64): 
    for trial in range(10):
         // Time the latency associated with reshaping the data to support a batch size of "batch_size"

You can do this locally without creating and deploying a container. Let me know if this is still unclear.

Reply:
@Corey-Zumar I'm saying don't do anything about the varying batch sizes for the user. Just provide them the model as they serialized it and provide them batches of inputs. This is what we do for every other model anyway. There are lots of other cases where the users have to account for variable-sized batches in their predict function code anyway. The only reason we extract out the mxnet model to be serialized separately is because it cannot be pickled. We are just trying to provide the user the ability to deploy predict functions that contain mxnet models.

I'm strongly against transparently modifying the model behind the users back. This makes their predict function harder to test and debug, and can make it harder to handle edge cases. E.g. what if they want to handle variable sized inputs as well as variable sized batches. I think we should maintain the contract we hold with other model deployers and simply stick to automatically serializing and deserializing their models.

Reply:
@dcrankshaw Awesome, that makes sense.